Responding to Issue #48 (#59)
uptane/deployment-considerations: faq.md

# Frequently asked questions

### **What makes Uptane different from other SOTA security mechanisms?**

> 

Reviewer: ```suggestion

Security problems occur due to accidental disclosures, malicious attacks, disgruntled insiders.  It is not
a matter of whether a successful attack will occur, but when.  One key feature of a security system is the
ability to securely recover from an attack.  This means that an update system must have a way to securely 
recover from a key loss or compromise.  

For example, suppose a nation-state actor steals a signing key and wants to use it to distribute software (as has
happened before) [cite Iran, etc.].  The update system must provide a way to revoke the current trusted
information even if the adversary is able to be a man-in-the-middle for future communications.  Uptane 
is designed to provide strong security in cases like these and is designed so that failures are 
compartmentalized and limited in scope.  

No other automotive update system has been designed to work in such rigorous situations or has 
received the public scrutiny of Uptane.  We follow best practice in the security community by having
wide-scale, public review.  This has been proven to be essential time and time again to ensure a 
design will hold up against attackers, especially those as strong as nation-state actors.  Furthermore,
Uptane's design is heavily influenced by the design of TUF, a widely used software update system 
with a strong track record of usability and security across millions of devices.  As a free and open
standard, with no cost to use or adopt, Uptane stands alone in the automotive update space.

```

Reviewer 2: Might be easier to review this if it were a separate PR. The linebreaks aren't helping, either.

### **How does Uptane work with other systems and protocols?**

> 

Reviewer: Other mechanisms for performing update such as XXX, YYY, ZZZ, are compatible with Uptane.  Uptane can use
any mechanism for data transport and is designed to provide strong security guarantees even if the 
underlying network or transport mechanism is compromised.  If an automaker wants to move to a secure
update system, keeping their existing update system as a transport for Uptane is an effective way to do so.

### **Must all signatures be valid for a threshold of signatures to be valid?**
The threshold requirement of Uptane, mentioned in [Section 5.4.4.3](https://uptane.github.io/uptane-standard/uptane-standard.html#check_root) and in descriptions of other roles, stipulates that a set number of keys are required to sign a metadata file. This is designed to prevent attacks by requiring would-be hackers to compromise multiple keys in order to install malware. However, what happens if this threshold is met, but one or more of the keys is not valid?

> As currently configured, the Uptane Standard allows each implementer to decide whether these keys would be accepted. Take the following example:

Reviewer 2: I would drop the "As currently configured" bit. I think that would just serve to confuse, because it sounds like we might change it, and I don't think we want to do that.

Author: Agree with Patrick's note

```suggestion
The Uptane Standard allows each implementer to decide whether these keys would be accepted. Take the following example:
```

> Root metadata says valid top-level targets keyids are A, B, C, and D, with a threshold of 2. Should the following two metadata files be considered valid?

Reviewer 2: ```suggestion
Root metadata lists valid top-level Targets key identifiers are A, B, C, and D, with a threshold of 2. Should the following two metadata files be considered valid?
```

> * Signed by A, B, and X, where X is not present at all in root metadata

Reviewer 2: ```suggestion
* Signed by A, B, and X, where X is not present at all in Root metadata
```

* Signed by A, B, and C, but B's signature doesn't actually match the signed content

> The first case can happen when you include X in a newer version of root metadata (for example the next iteration), so this has to be handled correctly or it will complicate the process of adding and rotating keys. The second case could happen when you are changing or adding signing algorithms. This case can occur if B is using a new signing scheme and the client does not notice, but will know how to parse it after the update.

Reviewer 2: ```suggestion
The first case can happen when you include X in a newer version of Root metadata (for example the next iteration), so this has to be handled correctly or it will complicate the process of adding and rotating keys. The second case could happen when you are changing or adding signing algorithms. This case can occur if B is using a new signing scheme that the client currently does not understand, but will know how to parse it after the update.
```

Author: Sorry for missing the capitalizations here. I'll look out for those moving forward.

> The first case can happen when you include X in a newer version of Root metadata (for example the next iteration), so this has to be handled correctly or it will complicate the process of adding and rotating keys. The second case could happen when you are changing or adding signing algorithms. This case can occur if B is using a new signing scheme that the client currently does not understand, but will know how to parse it after the update.

Reviewer 2: Sorry for another round. Rereading it, the "it" seems wrong and unnecessary to me.
```suggestion
The first case can happen when you include X in a newer version of Root metadata (for example the next iteration), so this has to be handled correctly or it will complicate the process of adding and rotating keys. The second case could happen when you are changing or adding signing algorithms. This case can occur if B is using a new signing scheme that the client currently does not understand, but will know how to parse after the update.
```

Author: I re-read this sentence several times because it did seem a little weird. Removing the "it" does make it flow a bit better.
